go/pkg/io/impath/impath.go

```diff
@@ -0,0 +1,152 @@
+// Package impath (immutable path) provides immutable and distinguishable types for absolute and relative paths.
+// This allows for string guarantees at compile time, improves legibility and reduces maintenance burden.
+package impath
+
+import (
+	"errors"
+	"fmt"
+	"path/filepath"
+	"strings"
+)
+
+var (
+	ErrNotAbsolute   = errors.New("impath: path is not absolute")
+	ErrNotRelative   = errors.New("impath: path is not relative")
+	ErrNotDescendant = errors.New("impath: target is not descendant of base")
+)
+
+type path string
+
+type Abs struct {
+	path
+}
+
+type Rel struct {
+	path
+}
+
+func (p path) String() string {
+	return string(p)
+}
+
+func (p path) dir() path {
+	return path(filepath.Dir(string(p)))
+}
+
+func (p path) base() path {
+	return path(filepath.Base(string(p)))
+}
+
+func (p Abs) Dir() Abs {
+	return Abs{path: p.dir()}
+}
+
+func (p Rel) Dir() Rel {
+	return Rel{path: p.dir()}
+}
+
+func (p Abs) Base() Abs {
+	return Abs{path: p.base()}
+}
+
+func (p Rel) Base() Rel {
+	return Rel{path: p.base()}
+}
+
+var (
+	zeroAbs = Abs{}
+	zeroRel = Rel{}
+)
+
+// ToAbs creates a new absolute and clean path from the specified parts.
+//
+// If the specified parts do not join to a valid absolute path, ErrNotAbsolute is returned.
+// If error is not nil, the returned value is not valid.
+func ToAbs(pathParts ...string) (Abs, error) {
+	p := filepath.Join(pathParts...)
+	if filepath.IsAbs(p) {
+		return Abs{path: path(p)}, nil
+	}
+	return zeroAbs, ErrNotAbsolute
+}
+
+// MustAbs is a convenient wrapper of ToAbs that is useful for constant paths.
+//
+// If the specified parts do not join to a valid absolute path, it panics.
+func MustAbs(pathParts ...string) Abs {
+	p, err := ToAbs(pathParts...)
+	if err != nil {
+		panic(err)
+	}
+	return p
+}
+
+// ToRel creates a new relative and clean path from the specified parts.
+//
+// If the specified parts do not join to a valid absolute path, ErrNotRelative is returned.
+// If error is not nil, the returned value is not valid.
+func ToRel(pathParts ...string) (Rel, error) {
+	p := filepath.Join(pathParts...)
+	if filepath.IsAbs(p) {
+		return zeroRel, ErrNotRelative
+	}
+	return Rel{path: path(p)}, nil
+}
+
+// MustRel is a convenient wrapper of ToRel that is useful for constant paths.
+//
+// If the specified parts do not join to a valid relative path, it pacnis.
+func MustRel(pathParts ...string) Rel {
+	p, err := ToRel(pathParts...)
+	if err != nil {
+		panic(err)
+	}
+	return p
+}
+
+// JoinAbs is a convenient method to join multiple paths into an absolute path.
+//
+// If the specified arguments do not join to a valid absolute path, ErrNotAbsolute is returned.
+// This happens if the parent is a zero path (not created using a function from this package).
+// If error is not nil, the returned value is not valid.
+func JoinAbs(parent Abs, parts ...Rel) (Abs, error) {
```

Comment:
The name suggests joining Abs might be misleading. Might be clearer to name it `AppendToAbs` or even make it a member function (called sth like `Append`) of of `Abs` type.

Reply:
Good point. I have eyed this name multiple times. `Join` does suggest that all arguments are of the same type, which is not ideal in this case. I was hesitant to define it as a member since it doesn't depend on any state. I made an exception above for `Dir` and `Base` to maintain consistent naming between the two types. I guess it makes sense to make the same exception here where `Append` would have the same argument list for both types.

Reply:
Replaced this with an `Append` method that has identical signature for both types.